games/xworld3d/x3item.h

```diff
@@ -0,0 +1,265 @@
+
+    X3ItemInfo(std::string n, X3EntityType t, std::string f,
+               const Loc3& l) :
+            name(n), type(t), model_file(f), loc(l) {}
+
+    X3ItemInfo(std::string n, X3EntityType t, std::string f,
+               const std::vector<int>& v) :
+            name(n), type(t), model_file(f), loc(v) {}
+
+    std::string name;
+    X3EntityType type;
+    std::string model_file;
+    Loc3 loc;
+};
+
+class X3Item : public X3Entity {
+public:
+    X3Item(const X3ItemInfo& info, World& world);
+
+    virtual ~X3Item() {
+        this->destroy();
+    }
+
+    const Object& object() {
+        return object_;
+    }
+
+    Pose pose() { return object_.pose(); }
+
+    void set_pose(const Pose& pose);
+
+    void set_speed(double vx, double vy, double vz);
+
+    void set_pose_and_speed(const Pose& pose, double vx, double vy, double vz);
+
+    bool equal(const X3Item* i) {
+        return (name_ == i->name());
+    }
+
+    void destroy() override;
+
+protected:
+    std::string model_file_;
+    Object object_;
+};
+
+/*********************************** X3Block **********************************/
+class X3Block : public X3Item {
+public:
+    X3Block(const X3ItemInfo& info, World& world) : X3Item(info, world) {
+        CHECK(info.type == X3EntityType::BLOCK);
+    }
+};
+
+typedef std::shared_ptr<X3Block> X3BlockPtr;
+
+/*********************************** X3Goal ***********************************/
+class X3Goal : public X3Item {
+public:
+    X3Goal(const X3ItemInfo& info, World& world) : X3Item(info, world) {
+        CHECK(info.type == X3EntityType::GOAL);
+    }
+};
+
+typedef std::shared_ptr<X3Goal> X3GoalPtr;
+
+/*********************************** X3Agent **********************************/
+class X3Agent : public X3Item {
+public:
+    X3Agent(const X3ItemInfo& info, World& world,
+            float speed_norm, int orientation_bins,
+            float reaching_dist);
+
+    void move_forward();
+
+    void turn_left();
+
+    void turn_right();
+
+    void jump();
+
+    void get_direction(double &dir_x, double &dir_y) {
+        dir_x = dir_x_;
+        dir_y = dir_y_;
+    }
+
+    float reach_test(const Pose& gpose);
+
+private:
+    void set_direction();
+
+    const float speed_norm_;
+    int yaw_id_;
+    double dir_x_;
+    double dir_y_;
+    const int orientation_bins_;
+    const float reaching_dist_; // An agent can collect this goal if it is
+                                // within the reaching distance of this goal
+};
+
+typedef std::shared_ptr<X3Agent> X3AgentPtr;
+
+/********************************** X3Camera **********************************/
+class X3Camera {
+public:
+    X3Camera(World& world, int img_height, int img_width);
+
+    Pose pose() { return camera_.pose(); }
+
+    roboschool::RenderResult render(X3Agent* agent, bool debug = false);
+
+    void detach() { agent_ = NULL; }
+
+private:
+    void attach_agent(X3Agent* agent);
+
+    void update(bool is_debug);
```

Comment:
is_debug and debug in line 243, either change to better names or add explanation on their functionality.